File: picker/recommender.py
"""Turns this week's parsed ESPN games into ranked survivor-pool pick candidates.

This module only produces output for a human to read -- it never submits
anything to the pool. It is deliberately simple: rank each team not yet
used by an entry by its estimated win probability (falling back to the
betting spread when ESPN's probability endpoint has no data yet), and
flag when both entries would end up recommended the same team so you can
diversify if you want to.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional

from data.models import Game
from models.win_prob import resolve_team_win_probability


@dataclass
class PickCandidate:
    team_abbreviation: str
    team_name: Optional[str]
    opponent_abbreviation: Optional[str]
    win_pct: Optional[float]  # 0-100, may be estimated
    win_pct_is_estimated: bool
    spread_detail: Optional[str]
    event_id: Optional[str]


def _candidates_for_game(game: Game) -> List[PickCandidate]:
    candidates = []
    spread_detail = game.odds.details if game.odds else None

    for team, opponent, is_home in (
        (game.home, game.away, True),
        (game.away, game.home, False),
    ):
        if not team.abbreviation:
            continue
        # Always 0-100 scale here, whether it came from ESPN's own
        # probability field or a spread-derived fallback -- see
        # models/win_prob.py for the blending logic.
        resolved = resolve_team_win_probability(game, is_home)
        candidates.append(
            PickCandidate(
                team_abbreviation=team.abbreviation,
                team_name=team.display_name,
                opponent_abbreviation=opponent.abbreviation,
                win_pct=resolved.win_pct,
                win_pct_is_estimated=resolved.source == "spread_estimate",
                spread_detail=spread_detail,
                event_id=game.event_id,
            )
        )
    return candidates


def rank_candidates(games: List[Game], used_teams: List[str]) -> List[PickCandidate]:
    """Rank all not-yet-used teams playing this week by estimated win probability.

    Games/teams with no probability data at all (win_pct is None) sort last,
    since we have no basis to recommend them over a team we can score.
    """
    all_candidates: List[PickCandidate] = []
    for game in games:
        if game.state and game.state != "pre":
            # Already started/finished games aren't legitimate picks for this week.
            continue
        all_candidates.extend(_candidates_for_game(game))

    available = [c for c in all_candidates if c.team_abbreviation not in used_teams]
    available.sort(key=lambda c: (c.win_pct is None, -(c.win_pct or 0)))
    return available


def recommend_for_entries(
    games: List[Game], used_teams_by_entry: Dict[str, List[str]], top_n: int = 5
) -> Dict[str, List[PickCandidate]]:
    """Return the top ``top_n`` ranked candidates for each entry.

    Each entry is ranked independently against its own used-teams history,
    since Entry A and Entry B can have burned different teams in past weeks.
    """
    return {
        entry: rank_candidates(games, used_teams)[:top_n]
        for entry, used_teams in used_teams_by_entry.items()
    }


def find_conflicts(recommendations: Dict[str, List[PickCandidate]]) -> Optional[str]:
    """If every entry's #1 recommendation is the same team, return that team's
    abbreviation as a heads-up to diversify; otherwise return None.
    """
    top_picks = {
        entry: candidates[0].team_abbreviation
        for entry, candidates in recommendations.items()
        if candidates
    }
    values = list(top_picks.values())
    if len(values) > 1 and len(set(values)) == 1:
        return values[0]
    return None

File: deadpool/src/engine/strategies/recommender.js
/**
 * Rank every not-yet-used team by win probability. The simple one.
 *
 * Port of `picker/recommender.py`, unchanged. This is the only
 * strategy the Python CLI could actually run — `main.py` imports this module
 * and nothing else — so it is the behaviour anyone who has used the terminal
 * tool has actually seen, and it is here as the baseline the others are
 * measured against rather than as the default.
 *
 * It reasons about one entry at a time and knows nothing about the other, so
 * it will happily hand both entries the same team. `findConflicts` is the
 * whole of its answer to that: a heads-up, after the fact.
 */

import { buildOptions, notUsed, byWinPctDesc, modelFieldsOf } from '../constraints.js';
import { f1 } from '../fmt.js';

export const ID = 'ranked';

const toCandidate = (o) => ({
  teamAbbreviation: o.teamAbbreviation,
  teamName: o.teamName,
  opponentAbbreviation: o.opponentAbbreviation,
  winPct: o.winPct,
  winPctIsEstimated: o.winPctSource === 'spread_estimate',
  winPctSource: o.winPctSource,
  spreadDetail: o.spreadDetail,
  eventId: o.eventId,
  startDate: o.startDate,
  ...modelFieldsOf(o),
});

/**
 * All not-yet-used teams playing this week, best first.
 *
 * Teams with no probability at all sort last rather than being dropped: there
 * is no basis to recommend one over a team we can score, and no basis to
 * pretend it is not playing either.
 */
export function rankCandidates(games, usedTeams, modelOpts = {}) {
  return notUsed(buildOptions(games, modelOpts), usedTeams)
    .map(toCandidate)
    .sort(byWinPctDesc);
}

/**
 * The top `topN` for each entry.
 *
 * Each entry ranks against its own used-teams history, because A and B can
 * have burned different teams in past weeks.
 */
export function recommendForEntries(games, usedTeamsByEntry, topN = 5) {
  const out = {};
  for (const [entry, used] of Object.entries(usedTeamsByEntry)) {
    out[entry] = rankCandidates(games, used).slice(0, topN);
  }
  return out;
}

/**
 * The team every entry's #1 lands on, if they all land on the same one.
 *
 * Null when the entries disagree, which is the healthy case.
 */
export function findConflicts(recommendations) {
  const tops = Object.values(recommendations)
    .filter((c) => c.length)
    .map((c) => c[0].teamAbbreviation);
  if (tops.length > 1 && new Set(tops).size === 1) return tops[0];
  return null;
}

const describe = (c) => {
  const winPct = c.winPct === null || c.winPct === undefined ? 'unknown' : `${f1(c.winPct)}%`;
  const basis = c.winPctIsEstimated ? ' (estimated from spread)' : '';
  const spread = c.spreadDetail ? `, spread ${c.spreadDetail}` : '';
  return `${c.teamAbbreviation} vs ${c.opponentAbbreviation || '?'} -- ${winPct} win prob${basis}${spread}`;
};

/* ------------------------------------------------ the registry contract -- */

export default {
  id: ID,
  name: 'Most likely to win',
  blurb: 'Takes the safest team on the board, out of the ones you have not used yet. '
    + 'Nothing else — no thought about later weeks, and no coordination between your two entries.',
  entries: 'single',
  params: [
    { key: 'topN', label: 'Alternatives to keep', type: 'int', default: 5, min: 1, max: 32, help: 'How far down the ranking to carry into the interface.' },
  ],

  run(ctx) {
    const topN = ctx.params.topN ?? 5;
    const perEntry = {};
    const picks = [];

    for (const entry of ctx.entries) {
      const ranked = rankCandidates(ctx.games, ctx.usedTeams[entry.id] ?? [], ctx.modelOpts);
      perEntry[entry.id] = ranked.slice(0, topN);
      const top = ranked[0] ?? null;
      picks.push({
        entry: entry.id,
        candidate: top,
        factors: top ? [
          { label: 'Win probability', value: top.winPct === null ? null : `${f1(top.winPct)}%`, weight: 1, note: top.winPctIsEstimated ? 'Estimated from the spread — ESPN has not published a probability yet.' : 'From ESPN.' },
          { label: 'Spread', value: top.spreadDetail ?? null, weight: 0 },
        ] : [],
        reasoning: top ? `Top pick: ${describe(top)}.` : 'No eligible teams available this week (all used, or no game data).',
      });
    }

    const conflict = findConflicts(perEntry);
    return {
      strategyId: ID,
      picks,
      candidates: perEntry,
      considered: Object.values(perEntry).reduce((n, c) => n + c.length, 0),
      warnings: conflict
        ? [{ level: 'warn', text: `Both entries' top pick is ${conflict}. One result would eliminate both — consider taking the runner-up for one of them.` }]
        : [],
    };
  },
};
